### ISSUE TITLE
rewrite latex editor to have a stateful backend component (was sagetex: use lock for exclusive operations)

### ISSUE THREAD

**Issue by haraldschilly** [sagemathinc/cocalc#422]
sometimes, when working on tex files longer lasting sagetex tasks pile up. use flock or something similar to be able to terminate the previous run and start the next one immediately without piling up a bunch of sagetex runs.

**Comment by williamstein** [comment]
The current latex editor is 100% client side, not using the server at all.    The right approach is to have a single backend object that completely manages running latex across all users who might be editing the document.     My sync2 branch redesigns the architecture to support this, but I haven't implemented it specifically for latex yet.

**Comment by haraldschilly** [comment]
Hmm, I specifically just meant this "sagetex" step, where an extracted sage/python file is executed via sage. Having a manager task, or some singleton object handing this, could also take care of that.

**Comment by billpage** [comment]
Alternatively, perhaps one could disable the automatic preview and introduce a new button to manually initiate preview.

**Comment by williamstein** [comment]
Preview is triggered by save.  So you could temporarily increase your
autosave interval in account settings to test this out.

On Tuesday, July 5, 2016, Bill Page <notifications@github.com> wrote:

> Alternatively, perhaps one could disable the automatic preview and
> introduce a new button to manually initiate preview.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/sagemathinc/smc/issues/422#issuecomment-230469331>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABN5dok8hF9veAAAm1Pxa5oIB3D8ZNRZks5qSlOOgaJpZM4HbJd4>
> .
>


-- 
Sent from my massive iPhone 6 plus.

**Comment by williamstein** [comment]
I spent Oct-April rewriting realtime sync rewrite to create the foundation for properly implementing the latex editor, among other things, using similar techniques to how sage worksheets work.  The latex editor can definitely be properly implemented in a way that is at least as powerful as Sage worksheets… so will properly respect multiple simultaneous editors, long run times, etc.   

>  Having a manager task, or some singleton object handing this, could also take care of that.

That's pretty much what I mean, where the object is just part of the local hub.   This is how sage worksheets work now.

**Comment by billpage** [comment]
Thanks William.  I have set autosave interval to 900 seconds.  Apparently
this is an "account" setting and not a project or file setting, right?  And
I guess 900 seconds is the maximum?  Is there some way to disable it
entirely?  If so, then at least in a single-user environment Save would
serve to trigger the Preview.

This seems to make things more manageable but I still have questions.  For
example in the Build tab log output I see

...
Code block begin...end
Inline formula 18
Inline formula 19
Inline formula 20
Code block begin...end
Inline formula 21
Code block begin...




-----------------------------------------------------
Running latex...

...

It looks like LaTeX is run before the Sage script completed.  Is this
normal?

**Comment by williamstein** [comment]
On Tue, Jul 5, 2016 at 8:34 AM, Bill Page <notifications@github.com> wrote:

> Thanks William. I have set autosave interval to 900 seconds. Apparently
> this is an "account" setting and not a project or file setting, right?


Correct.


> And
> I guess 900 seconds is the maximum?


Correct.


> Is there some way to disable it
> entirely?


No.  And if another user happens to have the file open with a short
autosave interval, there will be trouble.


> If so, then at least in a single-user environment Save would
> serve to trigger the Preview.
>
> This seems to make things more manageable but I still have questions. For
> example in the Build tab log output I see
>
> ...
> Code block begin...end
> Inline formula 18
> Inline formula 19
> Inline formula 20
> Code block begin...end
> Inline formula 21
> Code block begin...
>
>
>
>
> -----------------------------------------------------
> Running latex...
>
> ...
>
> It looks like LaTeX is run before the Sage script completed. Is this
> normal?
>
>
All that happens is that the latex command listed in the box, which you can
edit, gets run.  You can manually run it yourself in a terminal and watch
what happens.   After latex completes, sage is then run.    There are
likely timeouts about how long anything can run, and if they are exceeded,
there will be problems.



>
> On 5 July 2016 at 10:38, William Stein <notifications@github.com> wrote:
>
> > Preview is triggered by save. So you could temporarily increase your
> > autosave interval in account settings to test this out.
> >
> > On Tuesday, July 5, 2016, Bill Page <notifications@github.com> wrote:
> >
> > > Alternatively, perhaps one could disable the automatic preview and
> > > introduce a new button to manually initiate preview.
> > >
> > > —
> > > You are receiving this because you commented.
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/sagemathinc/smc/issues/422#issuecomment-230469331
> >,
> > > or mute the thread
> > > <
> >
> https://github.com/notifications/unsubscribe/ABN5dok8hF9veAAAm1Pxa5oIB3D8ZNRZks5qSlOOgaJpZM4HbJd4
> > >
> > > .
> > >
> >
> >
> > --
> > Sent from my massive iPhone 6 plus.
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/sagemathinc/smc/issues/422#issuecomment-230497132>,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe/AEwqQM9Pb98U0Glj5_edx5Lwui8IO1HTks5qSmxNgaJpZM4HbJd4
> >
>
> > .
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/sagemathinc/smc/issues/422#issuecomment-230514332>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABN5dhX9MfIIfK1LFlHTdpDCn9S3Mny2ks5qSnl6gaJpZM4HbJd4>
> .
>



-- 
William (http://wstein.org)

**Comment by billpage** [comment]
The latex command that I see in the Build tab is

pdflatex -synctex=1 -interact=nonstopmode 'discretization.tex'

If I run that in a terminal session all I see is a single pass of the
pdflatex which presumably generates a new 'discretization.sagetex.sage'
file.  Sage itself is apparently not auto run by this command.  It is the
Sage step that takes a long time.

**Comment by williamstein** [comment]
On Tue, Jul 5, 2016 at 8:45 AM, Bill Page <notifications@github.com> wrote:

> The latex command that I see in the Build tab is
>
> pdflatex -synctex=1 -interact=nonstopmode 'discretization.tex'
>
> If I run that in a terminal session all I see is a single pass of the
> pdflatex which presumably generates a new 'discretization.sagetex.sage'
> file. Sage itself is apparently not auto run by this command. It is the
> Sage step that takes a long time.
>

The sage command is

   sage discretization.sagetex.sage

It is only run *because* the output of pdflatex says to run it.  You should
run it manually to see what happens.

> Sage itself is apparently not auto run by this command.

If it were, then the act of latex'ing any file on the internet or using any
package, could trivially delete all your files, install spyware, etc.  --
that would be a massive security issue.  (Just explaining why latex makes
things hard.... and why moving work into "containers" like SMC are so
awesome.)


>
>
> On 5 July 2016 at 11:38, William Stein <notifications@github.com> wrote:
>
> > On Tue, Jul 5, 2016 at 8:34 AM, Bill Page <notifications@github.com>
> > wrote:
> >
> > >
> > > This seems to make things more manageable but I still have questions.
> For
> > > example in the Build tab log output I see
> > >
> > > ...
> > > Code block begin...end
> > > Inline formula 18
> > > Inline formula 19
> > > Inline formula 20
> > > Code block begin...end
> > > Inline formula 21
> > > Code block begin...
> > >
> > >
> > >
> > >
> > > -----------------------------------------------------
> > > Running latex...
> > >
> > > ...
> > >
> > > It looks like LaTeX is run before the Sage script completed. Is this
> > > normal?
> > >
> > >
> > All that happens is that the latex command listed in the box, which you
> can
> > edit, gets run. You can manually run it yourself in a terminal and watch
> > what happens. After latex completes, sage is then run. There are
> > likely timeouts about how long anything can run, and if they are
> exceeded,
> > there will be problems.
> >
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/sagemathinc/smc/issues/422#issuecomment-230517774>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABN5dtPReKaJLE96ElfGVLNExFWsn0-eks5qSnwegaJpZM4HbJd4>
> .
>



-- 
William (http://wstein.org)

**Comment by haraldschilly** [comment]
@billpage there is a sequence of conditional commands hardcoded into the front-end code. you can look at it here: https://github.com/sagemathinc/smc-public/blob/master/page/editor.coffee#L2591

by the way, the timeout for running sage is 45 seconds, as you can see here in `_run_sage`: https://github.com/sagemathinc/smc-public/blob/master/page/editor.coffee#L2686

my initial idea was, to move this sequence of commands into a dedicated management utility in a python library running in the project, which uses `latexmk` (or when the command is specified, that command or something else), etc.  william's approach is to have some stateful management object, which I don't fully understand. maybe that's orthogonal to each other, not sure. 

however, long story short, in your case you just have to run `sage *.sagetex.sage` and wait until it finishes. the results of that are then cached for your next latex runs. which means, they shouldn't take such a long time, it's only the sage step which takes longer...

**Comment by billpage** [comment]
"timeout for running sage is 45 seconds"

Ah, that makes sense now.  I guess there is no way to change that right now.

Anyway, I am still a bit confused by what happens when I edit the command
in the Build tab ...

With the default value I do see a sequence of commands run, i.e.
pdflatex;sage;pdflatex;pdflatex.  As I understand it (It is welll
documented in various SageTex documentation.), the first command generates
the .sage file, then sage runs and generates an .out file which the next
then pdflatex command reads to populate the sage output and then there is a
final pdflatex step for bibliography citations.

But it seems that if I modify this command then something else happens? For
example if I enter

   echo hi; pdflatex -synctex=1 -interact=nonstopmode 'discretization.tex'

in the box and press enter so that I get

  %sagemathcloud={"latex_command":"echo hi; pdflatex -synctex=1
-interact=nonstopmode 'discretization.tex'"}

after the /end{document}.

Now if I make a change to the document and click Save, in the Build log
output I see


-----------------------------------------------------
Running latex...




hi
This is pdfTeX, Version 3.14159265-2.6-1.40.16 (TeX Live 2015/Debian)
(preloaded format=pdflatex)
 restricted \write18 enabled.
entering extended mode
(./discretization.tex
LaTeX2e <2015/01/01>
Babel <3.9l> and hyphenation patterns for 79 languages loaded.
(/usr/share/texlive/texmf-dist/tex/latex/base/article.cls
Document Class: article 2014/09/29 v1.4h Standard LaTeX document class
(/usr/share/texlive/texmf-dist/tex/latex/base/size10.clo))
(/usr/share/texlive/texmf-dist/tex/latex/titling/titling.sty)

...

but **this happens only once** and Sage is not run at all ???

So I decided to try entering the sequence of commands itself as the Build
command

  %sagemathcloud={"latex_command":"pdflatex -synctex=1
-interact=nonstopmode 'discretization.tex'; sage
'discretization.sagetex.sage'; pdflatex -synctex=1 -interact=nonstopmode
'discretization.tex'; pdflatex -synctex=1 -interact=nonstopmode
'discretization.tex'"}

This did run pdflatex followed immediately by Sage but unfortunately it
apparently timed out even sooner before completing the Sage step.

So what I have as a work-round right now is to go the Build tab, modify the
command a little, and press Enter in the command box so the modified
default latex command gets added to the end of my LaTex file., e.g.

  %sagemathcloud={"latex_command":"echo run once; pdflatex -synctex=1
-interact=nonstopmode 'discretization.tex'"}

Now when I edit the file the preview step *only* runs pdflatex once and
does not try to run Sage. This lets me (and presumably also my many online
collaborators ...) see my LaTeX changes almost immediately. Sometimes I
have to click the Preview button before the preview is updated.

Now I also setup a separate terminal session with the command

  ~$ sage 'discretization.sagetex.sage'

ready to run. When I want to see updated Sage output in the preview I just
execute that command and return to the editor to see the updated preview.
It's a bit awkward but this is working OK for me right now.


Bill Page.

**Comment by billpage** [comment]
Oops, not so good after all. For reasons I don't understand, the behaviour
changed and now the editor tries to run Sage anyway. Maybe this happens
only if I make an edit that would change the generated file that will be
passed to sage.  Although there is a time-out of 45 seconds and the
sequence of commands continues with the next pdflatex command, the sage
continues to run until it completes. :(

The only thing that seems to really work is what William suggested: change
the Build command to something like 'echo' that does nothing, then run
pdflatex (and optionally sage) in a terminal session.

**Comment by haraldschilly** [comment]
> For reasons I don't understand, the behaviour changed and now the editor tries to run Sage anyway.

well, no and yes, it should always be the case, that the first build step runs your latex command, and then it decides based on the output, what it should do next. if sagetex has done something that causes a change, it says something like "please run sage again" and then SMC does that for you. only by stopping this first build command from running latex, you can be sure that this never happens. that's why "echo" comes into play here.

and yes, that 45 secs is a timeout from the client, and it seems it isn't really caring much about the still running sage instance.

your second last comment also shows, that this is not so trivial to do in practice and automatically -- that's what I looked at earlier based on your comment. i.e. it would not make much sense, to always stop everything (latex, bibtex and sage) from continuing running, when there is an edit. in the worst case, you'll never get an updated pdf while you keep insisting to render the preview. Just for the sage step, it would be better to wait up to a few minutes to run (but only run it once at the same time!), while at the same time several latex build steps can be run still produce a rapid preview. The sagetex results are cached, and hence it is possible to do this in parallel. (only caveat, the outputs are numbered ... e.g. in such a case where you insert a new sagetex command in between two others, the association between input and output will off by one after that)

**Comment by billpage** [comment]
I have a simple hack that implements the original suggestion to kill
old long running sage processes before starting new ones. I wrote a
small script named 'sage' in my local project bin that kill's any
running sage process with same file name before starting a new one
with that name.

~$ cat ~/bin/sage

#! /bin/bash
if [ "$1" != "" ]; then
  echo cleanup: $1 `eval "ps -Af | grep 'python $1' | grep -v grep |
awk '{print "'$2'"}'"`
  kill `eval "ps -Af | grep 'python $1' | grep -v grep | awk '{print "'$2'"}'"`
fi
/usr/local/bin/sage $*


This script will be called by the SMC tex editor when it tries to
build a new preview.  This way I can avoid the problem of
unintentionally having more than one sage process running at a time
due to processing delays in the SMC tex editor.

**Comment by williamstein** [comment]
Make sure to clear remotely cached files properly -- see https://github.com/sagemathinc/smc/issues/354

**Comment by williamstein** [comment]
(small vague note): At dinner tonight @haraldschilly  pointed out that one can easily do this by just emulating what is done with video chat, namely: use the embedded latex command configuration to store and share/sync the relevant "I'm now compiling" state.   This isn't quite 100% though, due to race conditions -- if two users try to tex at once it'll still run both of them, so really *something* has to happen on the backend...

**Comment by haraldschilly** [comment]
I'm closing this ticket. All the above is handled differently now. Whatever issue is left after the rewrite, it is better clarified in a new ticket.
